### PR TITLE
Introduce minimal Schema for nanopore registration 

### PR DESCRIPTION
**Pull request by Steffengreiner**
**What was changed**
This PR introduces a schema and example json for checking if the minimal required files for a nanopore data registration are within a provided dataset. 
Additionally it adapts the `OxfordNanoporeExperiment` generation to not fail if an unknown file or folder is found but include them as `OptionalFile` and `OptionalFolder` in the datastructure to allow for variations within the provided datastructure.

**Background information**
As of now the minimum required folders are: 

- FASTQ_FAIL
- FASTQ_PASS
- FAST5_FAIL
- FAST5_PASS

While the minimum required files are: 

- final_summary.txt (We extract metadata from this file)
- report.md (We extract metadata from this file)
- sequencing_summary.txt (Both Labs)

**Open Questions**
Due to the new constraints, the previously defined schemas and datafiles/Datafolders are ALL now unnecessary and could be deleted, since as soon as the minimal required files are provided the `OxfordNanoporeExperiment` is created and can be parsed.
The only reason for keeping multiple schemas could be if the distinction between the files in the OxfordnanoporeExperiment class is necessary (e.g. for accessing metadata within the files).
However this is only the case for the `final_summary.txt` file and the `report.md` file as of yet.

### CODE REVIEW

File: src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy [hunk 1]
```diff
@@ -14,239 +16,258 @@ import java.lang.reflect.Method
...
-            throw new IllegalArgumentException("No QBiC sample identifier found!")
+  /*
+   * Helper method that creates a list of mixed DataFolders and DataFiles instances
+   */
+
+  private static List<?> parseMeasurementItems(List<Map> items) {
+    final def children = []
+    items.each { item ->
+      {
+        if (item.get(FILEKEY)) {
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy:137]: Maybe extract this into a method `isFile():boolean` and make the null or empty check more expressive. It took me a while to understand that you check whether the item has a property called `file_type` provided by `FILEKEY` you check for here.

File: src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy [hunk 2]
```diff
@@ -14,239 +16,258 @@ import java.lang.reflect.Method
...
+          // Lets try to parse it as a subclass of a DataFile
+          DataFile putativeFile = parseFile(item)
+          children.add(putativeFile)
         }
-        if (ids.size() > 1) {
-            throw new IllegalArgumentException("Name contained more than one valid sample id!")
+        if (item.get(FOLDERKEY)) {
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy:142]: Same as with the `FILEKEY`, I think this is not very easy to read and understand.

File: src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy [hunk 3]
```diff
@@ -14,239 +16,258 @@ import java.lang.reflect.Method
...
-        final def ids = SampleCodeFunctions.findAllQbicSampleCodes(name as String)
-        if (ids.isEmpty()) {
-            throw new IllegalArgumentException("No QBiC sample identifier found!")
+  /*
+   * Helper method that creates a list of mixed DataFolders and DataFiles instances
+   */
+
+  private static List<?> parseMeasurementItems(List<Map> items) {
+    final def children = []
+    items.each { item ->
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy:135]: Why did you change from `IllegalArgumentException` to the differentiation by properties present?

File: src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy [hunk 4]
```diff
@@ -14,239 +16,258 @@ import java.lang.reflect.Method
...
+    def parsedChildren = []
+    children.each { Map unknownChild ->
+      if (unknownChild.get(FILEKEY)) {
+        def child = parseFile(unknownChild)
+        parsedChildren.add(child)
+      }
+      if (unknownChild.get(FOLDERKEY)) {
+        // We do not capture the second parse call, as we want to fail the parsing at this point.
+        // This means that we ultimately found a child of unknown type, which should
+        // break the parsing.
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy:250]: Is this still valid?

File: src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy [hunk 5]
```diff
@@ -14,239 +16,258 @@ import java.lang.reflect.Method
...
+        def child = parseFolder(unknownChild)
+        parsedChildren.add(child)
+      }
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy:253]: reminds me of the measurement item parsing. Please refer to my comments there.

**Review by KochTobi — CHANGES_REQUESTED** [review]
Only some minor requests

File: src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy (hunk 4 above)
Comment (Steffengreiner) [reply src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy:250]: Nice catch 👀

File: src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy [hunk 6]
```diff
@@ -14,239 +16,258 @@ import java.lang.reflect.Method
...
-            FQDN_FOLDERS + ".OtherReportsFolder",
-            FQDN_FOLDERS + ".BasecallingFolder"
-    ]
-
-    private OxfordNanoporeExperiment(String sampleId, List<OxfordNanoporeMeasurement> measurements) {
-        this.measurements = Objects.requireNonNull(measurements, "measurements must not be null")
-        this.sampleId = Objects.requireNonNull(sampleId, "sampleId must not be null")
-    }
+  // Fully qualified domain name of the nanopore folder structure package
+  private final static String FQDN_FOLDERS = "life.qbic.datamodel.datasets.datastructure.folders.nanopore"
```
[the remark this answers is not in the document]
Comment (Steffengreiner) [reply src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy]: Nice catch 👀

File: src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy [hunk 7]
```diff
@@ -14,239 +16,258 @@ import java.lang.reflect.Method
...
-     * @param Map nanoPoreSequencerOutput
-     * @return OxfordNanoporeExperiment A new instance of a nanopore experiment.
-     */
-    static OxfordNanoporeExperiment create(Map nanoPoreSequencerOutput) {
-        final String sampleId = parseQbicIdFromRootFolder(nanoPoreSequencerOutput)
-        final List<OxfordNanoporeMeasurement> measurements = parseMeasurements(nanoPoreSequencerOutput)
-        return new OxfordNanoporeExperiment(sampleId, measurements)
-    }
+  //Unique key in fileTreeMap identifying a child as a file
+  private final static String FILEKEY = "file_type"
```
[the remark this answers is not in the document]
Comment (Steffengreiner) [reply src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy]: Nice catch 👀

File: src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy [hunk 8]
```diff
@@ -14,239 +16,258 @@ import java.lang.reflect.Method
...
+  //Unique key in fileTreeMap identifyng a child as a folder
+  private final static String FOLDERKEY = "children"
+
+  private final List<OxfordNanoporeMeasurement> measurements
+
+  private final String sampleId
+
+  private final static Set nanoporeFileTypes = [
```
[the remark this answers is not in the document]
Comment (Steffengreiner) [reply src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy:33]: Nice catch 👀

File: src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy [hunk 9]
```diff
@@ -14,239 +16,258 @@ import java.lang.reflect.Method
...
-     */
-    static OxfordNanoporeExperiment create(Map nanoPoreSequencerOutput) {
-        final String sampleId = parseQbicIdFromRootFolder(nanoPoreSequencerOutput)
-        final List<OxfordNanoporeMeasurement> measurements = parseMeasurements(nanoPoreSequencerOutput)
-        return new OxfordNanoporeExperiment(sampleId, measurements)
-    }
+  //Unique key in fileTreeMap identifying a child as a file
+  private final static String FILEKEY = "file_type"
+  //Unique key in fileTreeMap identifyng a child as a folder
+  private final static String FOLDERKEY = "children"
```
[the remark this answers is not in the document]
Comment (Steffengreiner) [reply src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy]: Nice catch 👀

File: src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy [hunk 10]
```diff
@@ -14,239 +16,258 @@ import java.lang.reflect.Method
...
-    ]
-
-    private OxfordNanoporeExperiment(String sampleId, List<OxfordNanoporeMeasurement> measurements) {
-        this.measurements = Objects.requireNonNull(measurements, "measurements must not be null")
-        this.sampleId = Objects.requireNonNull(sampleId, "sampleId must not be null")
-    }
+  // Fully qualified domain name of the nanopore folder structure package
+  private final static String FQDN_FOLDERS = "life.qbic.datamodel.datasets.datastructure.folders.nanopore"
+  // Fully qualified domain name of the nanopore file structure package
+  private final static String FQDN_FILES = "life.qbic.datamodel.datasets.datastructure.files.nanopore"
```
[the remark this answers is not in the document]
Comment (Steffengreiner) [reply src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy]: Nice catch 👀

File: src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy [hunk 11]
```diff
@@ -14,239 +16,258 @@ import java.lang.reflect.Method
...
+          FQDN_FILES + ".ThroughputLog",
+          FQDN_FILES + ".BarcodeAlignmentLog",
+          FQDN_FILES + ".PoreActivityLog",
+          FQDN_FILES + ".SampleSheetLog",
+          FQDN_FILES + ".PoreScanDataLog",
+          FQDN_FILES + ".SequencingTelemetryLog",
+          FQDN_FILES + ".GuppyBasecallLog"
+  ]
+
+  private final static Set nanoporeFolderTypes = [
```
[the remark this answers is not in the document]
Comment (Steffengreiner) [reply src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy:55]: Good idea 🧠

File: src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy (hunk 1 above)
Comment (Steffengreiner) [reply src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy:137]: Good idea 🧠

File: src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy (hunk 2 above)
Comment (Steffengreiner) [reply src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy:142]: It's adapted nice idea 👍

File: src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy (hunk 5 above)
Comment (Steffengreiner) [reply src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy]: Will do 🥳

File: src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy (hunk 3 above)
Comment (Steffengreiner) [reply src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy:88]: Basically to avoid unnecessary exception generation since we now know what defines a file/folder before iterating over all possible combinations and allow for unknown items in the map. 

**In more detail:** 
Since we now allow for optional files and folders, parseFile() and parseFolder() will not throw an exception if the item is not recognized. Instead it will create an OptionalFile with the content of the provided item. 
However this is not possible if the item is a folder, since a folder does not have a file_type, which is required for the optional file generation. This leads to an exception which would have to be caught and redirected again. 

Also IMO It's more explicit since we explicitly show what the defining properties are since we rely on them during the File/Folder generation process anyway.

Comment (KochTobi) [reply src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy:88]: The original reasoning was that what makes a file a file and what makes a folder a folder is defined in the JSON schema. Thus only the method that uses the schema can determine whether we are indeed parsing a file or a folder. It first checks whether a file can be parsed, if not it tries to parse a folder. If neither works, then there should be an exception. The solution suggested in this PR does not consider that and instead hard-codes the properties.

Comment (KochTobi) [reply src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy:88]: As we have this logic in multiple places, perhaps it would be an option to extract the parsing or to restructure the code slightly.
If this is more complex than suspected or if I missed some aspects of it, I will not block this PR as it works.

**Review by KochTobi — APPROVED** [review]
LGTM :+1:

File: src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy (hunk 3 above)
Comment (Steffengreiner) [reply src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy:88]: I'd agree with you however the parseFile and parseFolder methods beforehand also expected specific properties to be there in the first place (namely "name", "path", and "children"). 
Additionally the parsing should not happen if the schema fails, therefore we can expect that these properties could be found during the validation process.

Comment (Steffengreiner) [reply src/main/groovy/life/qbic/datamodel/datasets/OxfordNanoporeExperiment.groovy:88]: I'd argue that down the line we should re-evaluate where the schema should be processed. 
IMO the strict seperation of schema and parsing does not make sense anymore and the schema should be automatically consulted if a NanoporeExperiment/NanoporeMeasurement is generated. 🤷‍♂️
